### PR TITLE
geo_shape support in index and query

### PR DESCRIPTION
**Pull request by will-r**
Implements the interface described in #765, with tests and readme. All geoJSON shapes are supported, mostly because we just pass them through without amendment or validation. There is a very little sugar for envelopes to give some consistency with geo_point searches, but elasticsearch isn't consistent in this area so perhaps that's not needed.

### CODE REVIEW

**Comment by ankane** [comment]
Hey, thanks @will-r! This is a great start. I'll add a few comments inline.

File: README.md [hunk 1]
```diff
@@ -834,7 +834,9 @@ product = Product.first
 product.similar(fields: ["name"], where: {size: "12 oz"})
 ```
 
-### Geospatial Searches
+### Simple Geospatial Indexing: geo_points
```
Comment (ankane) [inline README.md]: Let's leave this section as is.

File: README.md [hunk 2]
```diff
@@ -878,6 +882,81 @@ Also supports [additional options](https://www.elastic.co/guide/en/elasticsearch
 City.search "san", boost_by_distance: {field: :location, origin: {lat: 37, lon: -122}, function: :linear, scale: "30mi", decay: 0.5}
 ```
 
+
+### Complex Geospatial Indexing: geo_shapes
```
Comment (ankane) [inline README.md]: I think we can shorten this to `Geo Shapes`.

File: README.md [hunk 3]
```diff
@@ -878,6 +882,81 @@ Also supports [additional options](https://www.elastic.co/guide/en/elasticsearch
...
+Intersecting with the query shape:
+
+```ruby
+City.search "san", where: {bounds: {geo_shape: {type: "polygon", coordinates: [[{lat: 38, lon: -123}, ...]]}}}
+```
+
+Falling entirely within the query shape:
+
+```ruby
+City.search "san", where: {relation: "within", bounds: {geo_shape: {type: "circle", coordinates: [{lat: 38, lon: -123}], radius: "1km"}}}
```
Comment (ankane) [inline README.md]: It seems like `relation` should be inside the `geo_shape` hash.

```ruby
where: {bounds: {geo_shape: {type: "circle", coordinates: [{lat: 38, lon: -123}], radius: "1km", relation: "within"}}}
```

File: test/geo_shape_test.rb [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+class GeoShapeTest < Minitest::Test
+
+  def test_geo_shape
+    regions = [
+      {name: "Region A", text: "The witch had a cat", territory: "30,40,35,45,40,40,40,30,30,30,30,40"},
+      {name: "Region B", text: "and a very tall hat", territory: "50,60,55,65,60,60,60,50,50,50,50,60"},
+      {name: "Region C", text: "and long ginger hair which she wore in a plait.",  territory: "10,20,15,25,20,20,20,10,10,10,10,20"},
+    ]
+    store regions, Region
```
Comment (ankane) [inline test/geo_shape_test.rb]: It'd be nice to put region creation in a `setup` method, and split the assertions into separate tests - `test_circle`, `test_envelope`, etc.

File: test/test_helper.rb [hunk 5]
```diff
@@ -338,6 +368,33 @@ def search_routing
...
+  searchkick \
+    geo_shapes: {
+      territory: {tree: "quadtree", precision: "10km"}
+    }
+
+  def search_data
+    {
+      name: name,
+      text: text,
+      territory: as_geo_json
```
Comment (ankane) [inline test/test_helper.rb]: If we use a virtual attribute for `territory` (see how `aisle` field is done), we can pass an object during model creation and remove the methods below.

File: README.md [hunk 6]
```diff
@@ -878,6 +882,81 @@ Also supports [additional options](https://www.elastic.co/guide/en/elasticsearch
...
+```
+
+Not touching the query shape:
+
+```ruby
+City.search "san", where: {relation: "disjoint", bounds: {geo_shape: {type: "envelope", coordinates: [{lat: 38, lon: -123}, {lat: 37, lon: -122}]}}}
+```
+
+Envelope is a special case. For consistency, searchkick also understands coordinates given as top_left and bottom_right:
```
Comment (ankane) [inline README.md]: For simplicity, let's not special case this.

**Comment by ankane** [comment]
Had more comments than I expected, but overall, great job.

**Comment by will-r** [comment]
Thank you for taking the time to read through. It's much neater now, and I have finally noticed the destroy_alls in setup so Travis is happy.

**Comment by ankane** [comment]
Awesome, thanks @will-r. Looks great 💯

**Comment by ankane** [comment]
FYI renamed `geo_shapes` -> `geo_shape` as in the original proposal.
